File: packages/react-native-enriched-markdown/ios/renderer/LinkRenderer.m
#import "LinkRenderer.h"
#import "FontUtils.h"
#import "RenderContext.h"
#import "RendererFactory.h"
#import "StyleConfig.h"
#import <React/RCTFont.h>

@implementation LinkRenderer

#pragma mark - Rendering

- (void)renderNode:(MarkdownASTNode *)node into:(NSMutableAttributedString *)output context:(RenderContext *)context
{
  NSUInteger start = output.length;

  // 1. Render children first to establish base attributes
  [_rendererFactory renderChildrenOfNode:node into:output context:context];

  NSRange range = NSMakeRange(start, output.length - start);
  if (range.length == 0)
    return;

  // 2. Extract configuration
  NSString *url = node.attributes[@"url"] ?: @"";
  LinkVariantConfig *variant = [_config effectiveLinkVariantForURL:url];

  RCTUIColor *linkColor = variant.color ?: [_config linkColor];
  BOOL linkUnderline = variant ? variant.underline : [_config linkUnderline];
  NSString *linkFontFamily = [_config linkFontFamily];
  RCTUIColor *backgroundColor = variant ? variant.backgroundColor : [_config linkBackgroundColor];

  NSNumber *underlineStyle = @(linkUnderline ? NSUnderlineStyleSingle : NSUnderlineStyleNone);

  // 3. Apply core link functionality (non-destructive)
  [output addAttribute:NSLinkAttributeName value:url range:range];

  // 4. Optimize visual attributes via enumeration to avoid redundant updates
  [output enumerateAttributesInRange:range
                             options:NSAttributedStringEnumerationLongestEffectiveRangeNotRequired
                          usingBlock:^(NSDictionary<NSAttributedStringKey, id> *attrs, NSRange subrange, BOOL *stop) {
                            NSMutableDictionary *newAttributes = [NSMutableDictionary dictionary];

                            // Only apply link color if the subrange isn't already colored by the link style
                            if (linkColor && ![attrs[NSForegroundColorAttributeName] isEqual:linkColor]) {
                              newAttributes[NSForegroundColorAttributeName] = linkColor;
                              newAttributes[NSUnderlineColorAttributeName] = linkColor;
                            }

                            // Only update underline style if it differs from the config
                            if (![attrs[NSUnderlineStyleAttributeName] isEqual:underlineStyle]) {
                              newAttributes[NSUnderlineStyleAttributeName] = underlineStyle;
                            }

                            if (linkFontFamily.length > 0) {
                              UIFont *currentFont = attrs[NSFontAttributeName];
                              if (currentFont) {
                                UIFont *linkFont = [RCTFont updateFont:currentFont
                                                            withFamily:linkFontFamily
                                                                  size:nil
                                                                weight:nil
                                                                 style:nil
                                                               variant:nil
                                                       scaleMultiplier:1.0];
                                if (linkFont && ![currentFont isEqual:linkFont]) {
                                  newAttributes[NSFontAttributeName] = linkFont;
                                }
                              }
                            }

                            if (newAttributes.count > 0) {
                              [output addAttributes:newAttributes range:subrange];
                            }
                          }];

  if (backgroundColor) {
    [output addAttribute:NSBackgroundColorAttributeName value:backgroundColor range:range];
  }

  [context registerLinkRange:range url:url];
}

@end

File: packages/ios-enriched-markdown/Sources/EnrichedMarkdown/Rendering/Renderers/LinkRenderer.swift
import UIKit

final class LinkRenderer: NodeRenderer {
    private let factory: RendererFactory
    private let config: MarkdownStyleConfig

    init(factory: RendererFactory, config: MarkdownStyleConfig) {
        self.factory = factory
        self.config = config
    }

    func render(node: MarkdownASTNode, into output: NSMutableAttributedString, context: RenderContext) {
        let start = output.length
        factory.renderChildren(of: node, into: output, context: context)

        let range = RenderContext.rangeForRenderedContent(in: output, start: start)
        guard range.length > 0 else { return }

        let url = node.attribute("url") ?? ""
        let linkStyle = config.link
        let linkColor = linkStyle.foregroundColor
        let underline = linkStyle.underline ?? true
        let underlineStyle = underline ? NSUnderlineStyle.single.rawValue : 0

        let linkValue: Any = URL(string: url) ?? url
        output.addAttribute(.link, value: linkValue, range: range)

        output.enumerateAttributes(in: range, options: []) { _, subrange, _ in
            if let linkColor {
                output.addAttribute(.foregroundColor, value: linkColor, range: subrange)
                output.addAttribute(.underlineColor, value: linkColor, range: subrange)
            }
            output.addAttribute(.underlineStyle, value: underlineStyle, range: subrange)
        }
    }
}
